extensions-contrib/gce-extensions/src/main/java/org/apache/druid/indexing/overlord/autoscaling/gce/GCEAutoScaler.java

```diff
@@ -0,0 +1,496 @@
+                httpTransport,
+                jsonFactory
+        );
+        if (credential.createScopedRequired()) {
+          List<String> scopes = new ArrayList<>();
+          scopes.add(ComputeScopes.CLOUD_PLATFORM);
+          scopes.add(ComputeScopes.COMPUTE);
+          credential = credential.createScoped(scopes);
+        }
+
+        if (credential.getClientAuthentication() != null) {
+          log.error("Not using a service account, terminating");
+          System.exit(1);
+        }
+
+        cachedComputeService = new Compute.Builder(httpTransport, jsonFactory, credential)
+                .setApplicationName("DruidAutoscaler")
+                .build();
+
+        retries++;
+      }
+      catch (Throwable e) {
+        log.error(e, "Got Exception in creating the ComputeService");
+        throw e;
+      }
+    }
+    return cachedComputeService;
+  }
+
+  // Used to wait for an operation to finish
+  private Operation.Error waitForOperationEnd(
+      Compute compute,
+      Operation operation) throws Exception
+  {
+    final long pollInterval = 5 * 1000;  // 5 sec
+
+    String status = operation.getStatus();
+    String opId = operation.getName();
+    while (operation != null && !"DONE".equals(status)) {
+      Thread.sleep(pollInterval);
+      Compute.ZoneOperations.Get get = compute.zoneOperations().get(
+          envConfig.getProjectId(),
+          envConfig.getZoneName(),
+          opId
+      );
+      operation = get.execute();
+      if (operation != null) {
+        status = operation.getStatus();
+      }
+    }
+    return operation == null ? null : operation.getError();
+  }
+
+  /**
+   * When called resizes envConfig.getManagedInstanceGroupName() increasing it by creating
+   * envConfig.getNumInstances() new workers (unless the maximum is reached). Return the
+   * IDs of the workers created
+   */
+  @Override
+  public AutoScalingData provision()
+  {
+    final String project = envConfig.getProjectId();
+    final String zone = envConfig.getZoneName();
+    final int numInstances = envConfig.getNumInstances();
+    final String managedInstanceGroupName = envConfig.getManagedInstanceGroupName();
+
+    try {
+      List<String> before = getRunningInstances();
+      log.debug("Existing instances [%s]", String.join(",", before));
+
+      int toSize = Math.min(before.size() + numInstances, getMaxNumWorkers());
+      if (before.size() >= toSize) {
+        // nothing to scale
+        return new AutoScalingData(new ArrayList<>());
+      }
+      log.info("Asked to provision instances, will resize to %d", toSize);
+
+      Compute computeService = createComputeService();
+      Compute.InstanceGroupManagers.Resize request =
+              computeService.instanceGroupManagers().resize(project, zone,
+                      managedInstanceGroupName, toSize);
+
+      Operation response = request.execute();
+      Operation.Error err = waitForOperationEnd(computeService, response);
```

Comment:
In practice this has never been a big deal because the process of provisioning works in a loop calling the `provision` method and, as soon as a call returns an empty list, i.e., the case in which we get back a "DONE" but the machines are not up yet , the loop stops and at the next check for the need of machines they will be called again. 

With reference to the code of `PendingTaskBasedWorkerProvisioningStrategy` the loop works as follows:

```
while (workersToProvision > 0) {
  final AutoScalingData provisioned = workerConfig.getAutoScaler().provision();
  final List<String> newNodes;
  if (provisioned == null || (newNodes = provisioned.getNodeIds()).isEmpty()) {
    log.warn("NewNodes is empty, returning from provision loop");
    break;
  } else {
    ...
  }
}
```